## Pair Programming ANOVA





En el pair programming de hoy usaremos el set de datos que guardastéis en el pair programming de 
normalización y estandarización.


Hasta ahora habéis estado evaluando las características de vuestro set de datos y habéis hecho una 
gran exploración, es el momento de hacer vuestro primer ANOVA! En el ejercicio de hoy tendréis que 
hacer un ANOVA con vuestro datos y hacer una interpretación de los resultados.


📌 NOTA Puede que vuestros datos no se ajusten o no cumplan todas las asunciones, no pasa nada, 
haced el ANOVA e interpretad los resultados. En próximas lecciones aprenderemos que podemos hacer 
cuando nos encontramos en esta situación.

In [2]:
# Tratamiento de datos
# -----------------------------------------------------------------------
import numpy as np
import pandas as pd
import random 

# Gráficos
# ------------------------------------------------------------------------------
import matplotlib.pyplot as plt
import seaborn as sns

# Estadísticos
# ------------------------------------------------------------------------------
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.multivariate.manova import MANOVA
from sklearn.preprocessing import StandardScaler

plt.rcParams["figure.figsize"] = (10,8) 

In [3]:
df = pd.read_csv("datos/sephora_website_dataset3.csv", index_col = 0)
df.sample(5)

id               brand                  category  \
4381  1989706  Kiehl's Since 1851                 Face Wash   
4564  1947886       KORA Organics  Body Lotions & Body Oils   
5480   472845        Michael Kors                   Perfume   
1282  2114940            Caudalie              Moisturizers   
3022  2258515               Fresh                 Mini Size   

                                                   name             size  \
4381      Age Defender Dual-Action Exfoliating Cleanser     5 oz/ 150 mL   
4564                                 Noni Glow Body Oil  3.38 oz/ 100 mL   
5480                         Michael Kors Eau De Parfum           3.4 oz   
1282           Vinopure Natural Oil Control Moisturizer    1.3 oz/ 40 mL   
3022  Vitamin Nectar Glow Juice Antioxidant Face Ser...          no size   

      rating  number_of_reviews   love  price  value_price  ...  \
4381     5.0                  6   2000   28.0         28.0  ...   
4564     3.5                 38   8100   58.0         58.0  ...   
5480     4.5                510  10900  118.0        118.0  ...   
1282     3.5                 78   7500   39.0         39.0  ...   
3022     4.5                 41  18300   28.0         28.0  ...   

                                             how_to_use  \
4381  Suggested Usage:As a daily cleanser: \n-Dampen...   
4564  Suggested Usage:-Warm several drops in the pal...   
5480                                    no instructions   
1282  Suggested Usage:-In the morning and evening- a...   
3022  Suggested Usage:\n-Apply to cleansed skin dail...   

                                            ingredients online_only exclusive  \
4381   -Phytic Acid: Derived from rice bran and has ...           1         0   
4564  -Rosehip Oil: Provides nutrients and essential...           1         1   
5480  Alcohol Denat.- Michael Fragrance- Water- Lina...           1         0   
1282   -Grape Seed Polyphenols: Provides antioxidant...           0         0   
3022   -Vitamin Fruit Complex: A blend of vitamins C...           0         0   

     limited_edition limited_time_offer rating_norm  rating_log  rating_raiz  \
4381               0                  0    0.201996    1.609438     2.236068   
4564               0                  0   -0.098004    1.252763     1.870829   
5480               0                  0    0.101996    1.504077     2.121320   
1282               0                  0   -0.098004    1.252763     1.870829   
3022               0                  0    0.101996    1.504077     2.121320   

      rating_Sklearn  
4381             NaN  
4564             NaN  
5480             NaN  
1282             NaN  
3022             NaN  

[5 rows x 25 columns]

In [6]:
v_numericas = df.select_dtypes(include= ['int64', 'float64'] )

In [7]:
df_numericas = v_numericas[['rating', 'number_of_reviews', 'love', 'price', 'value_price']]

In [8]:
# iniciamos nuestro anova. Recordemos que la variable respuesta es rating
lm = ols('rating_norm ~ number_of_reviews + love + price + value_price' , data=df).fit()
sm.stats.anova_lm(lm)

df      sum_sq   mean_sq          F        PR(>F)
number_of_reviews     1.0    2.471909  2.471909  61.985612  3.853206e-15
love                  1.0    0.970796  0.970796  24.343699  8.200824e-07
price                 1.0    0.302955  0.302955   7.596901  5.858433e-03
value_price           1.0    3.198753  3.198753  80.211954  4.021957e-19
Residual           9163.0  365.409058  0.039879        NaN           NaN

In [9]:
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:            rating_norm   R-squared:                       0.019
Model:                            OLS   Adj. R-squared:                  0.018
Method:                 Least Squares   F-statistic:                     43.53
Date:                Tue, 09 May 2023   Prob (F-statistic):           3.01e-36
Time:                        10:33:23   Log-Likelihood:                 1762.9
No. Observations:                9168   AIC:                            -3516.
Df Residuals:                    9163   BIC:                            -3480.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
Intercept            -0.0131      0.003     -4.099      0.000      -0.019      -0.007
number_of_reviews  4.876e-06   3.52e-06      1.385      0.166   -2.02e-06    1.18e-05
love               3.748e-07   7.36e-08      5.094      0.000    2.31e-07    5.19e-07
price                 0.0023      0.000      9.309      0.000       0.002       0.003
value_price          -0.0021      0.000     -8.956      0.000      -0.003      -0.002
==============================================================================
Omnibus:                     5000.970   Durbin-Watson:                   1.744
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            33645.829
Skew:                          -2.621   Prob(JB):                         0.00
Kurtosis:                      10.785   Cond. No.                     7.01e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.01e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

R-squared: tiene que tener un valor entre 0 y 1, en el que 0 no explica la variabilidad de la variable respuesta en nuestro caso 0.019, y 1 indica que el modelo explica toda la variabilidad de la variable respuesta 

- El p value de love es 0.000 
- El coef 3.748e-07. 

CONCLUSIONES:

    - El valor del coefieciente es de 3.748e-07, por cada unidad que aumenta el love el rating disminuye un 0.000